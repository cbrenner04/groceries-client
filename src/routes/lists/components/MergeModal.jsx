import React from 'react';
import PropTypes from 'prop-types';
import { Button, Form, Modal } from 'react-bootstrap';

import { TextField } from '../../../components/FormFields';

function MergeModal({ showModal, clearModal, listNames, mergeName, handleMergeConfirm, handleMergeNameChange }) {
  return (
    <Modal show={showModal} onHide={clearModal}>
      <Modal.Header closeButton>
        <Modal.Title>Merge {`"${listNames}"`}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form>
          <TextField
            name="mergeName"
            label="Name for the merged list"
            value={mergeName}
            handleChange={handleMergeNameChange}
            placeholder="My super cool list"
          />
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="secondary" onClick={clearModal} data-test-id={`clear-merge`}>
          Close
        </Button>
        <Button
          variant="primary"
          onClick={handleMergeConfirm}
          data-test-id={`confirm-merge`}
          disabled={!mergeName}
          style={{ pointerEvents: mergeName ? 'auto' : 'none' }}
        >
          Merge lists
        </Button>
      </Modal.Footer>
    </Modal>
  );
}

MergeModal.propTypes = {
  showModal: PropTypes.bool.isRequired,
  clearModal: PropTypes.func.isRequired,
  listNames: PropTypes.string.isRequired,
  mergeName: PropTypes.string.isRequired,
  handleMergeConfirm: PropTypes.func.isRequired,
  handleMergeNameChange: PropTypes.func.isRequired,
};

export default MergeModal;
