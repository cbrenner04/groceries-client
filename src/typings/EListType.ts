const enum EListType {
  BOOK_LIST = 'BookList',
  GROCERY_LIST = 'GroceryList',
  MUSIC_LIST = 'MusicList',
  SIMPLE_LIST = 'SimpleList',
  TO_DO_LIST = 'ToDoList',
}

export default EListType;
