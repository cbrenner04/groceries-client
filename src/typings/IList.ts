import type EListType from './EListType';

export default interface IList {
  id?: string;
  name: string;
  type: EListType;
  created_at?: string;
  completed?: boolean;
  users_list_id?: string;
  owner_id?: string;
  refreshed?: boolean;
  categories?: string[];
}
