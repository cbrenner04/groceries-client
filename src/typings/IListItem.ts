export default interface IListItem {
  id?: string;
  product?: string;
  task?: string;
  content?: string;
  quantity?: string;
  author?: string;
  title?: string;
  artist?: string;
  album?: string;
  assignee_id?: string;
  due_by?: string;
  read?: boolean;
  number_in_series?: number;
  category?: string;
  completed?: boolean;
  purchased?: boolean;
  created_at?: string;
  user_id?: string;
  book_list_id?: string;
  grocery_list_id?: string;
  music_list_id?: string;
  simple_list_id?: string;
  to_do_list_id?: string;
}
