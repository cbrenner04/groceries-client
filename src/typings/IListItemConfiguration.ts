export default interface IListItemConfiguration {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  user_id: string;
  archived_at: string | null;
}
