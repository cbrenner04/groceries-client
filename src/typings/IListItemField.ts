export default interface IListItemField {
  id: string;
  list_item_field_configuration_id: string;
  data: string;
  archived_at: string;
  user_id: string;
  list_item_id: string;
  created_at: string;
  updated_at: string;
  label: string;
}
