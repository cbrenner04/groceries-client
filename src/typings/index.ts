export { default as EListType } from './EListType';
export { default as IListItemConfiguration } from './IListItemConfiguration';
export { default as IListItemField } from './IListItemField';
export { default as IListItemFieldConfiguration } from './IListItemFieldConfiguration';
export { default as IList } from './IList';
export { default as IListUser } from './IListUser';
export { default as IUsersList } from './IUsersList';
export { default as IListItem } from './IListItem';

// no idea why having this in its own file created type issues
export type TUserPermissions = Record<string, 'read' | 'write'>;
export enum EUserPermissions {
  WRITE = 'write',
  READ = 'read',
}
export enum EListItemFieldType {
  BOOLEAN = 'boolean',
  DATE_TIME = 'date_time',
  FREE_TEXT = 'free_text',
  NUMBER = 'number',
}
