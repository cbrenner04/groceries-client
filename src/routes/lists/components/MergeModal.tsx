import React, { ChangeEventHandler, MouseEventHandler } from 'react';
import { Button, Form, Modal } from 'react-bootstrap';

import { TextField } from '../../../components/FormFields';

interface IMergeModalProps {
  showModal: boolean;
  clearModal: () => void;
  listNames: string;
  mergeName: string;
  handleMergeConfirm: MouseEventHandler;
  handleMergeNameChange: ChangeEventHandler;
}

const MergeModal: React.FC<IMergeModalProps> = ({
  showModal,
  clearModal,
  listNames,
  mergeName,
  handleMergeConfirm,
  handleMergeNameChange,
}) => {
  return (
    <Modal show={showModal} onHide={clearModal}>
      <Modal.Header closeButton>
        <Modal.Title>Merge {`"${listNames}"`}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form>
          <TextField
            name="mergeName"
            label="Name for the merged list"
            value={mergeName}
            handleChange={handleMergeNameChange}
            placeholder="My super cool list"
          />
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="secondary" onClick={clearModal} data-test-id={`clear-merge`}>
          Close
        </Button>
        <Button
          variant="primary"
          onClick={handleMergeConfirm}
          data-test-id={`confirm-merge`}
          disabled={!mergeName}
          className={mergeName ? 'merge-modal-confirm-enabled' : 'merge-modal-confirm-disabled'}
        >
          Merge lists
        </Button>
      </Modal.Footer>
    </Modal>
  );
};

export default MergeModal;
