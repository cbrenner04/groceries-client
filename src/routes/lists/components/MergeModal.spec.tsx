import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MergeModal from './MergeModal';
import { EListType } from 'typings';

const defaultProps = {
  showModal: true,
  clearModal: jest.fn(),
  listNames: 'name, name',
  mergeName: '',
  handleMergeConfirm: jest.fn(),
  handleMergeNameChange: jest.fn(),
  selectedLists: [
    { id: '1', name: 'Test List 1', type: EListType.GROCERY_LIST },
    { id: '2', name: 'Test List 2', type: EListType.GROCERY_LIST },
  ],
};

function setup(suppliedProps = {}): ReturnType<typeof render> & {
  props: typeof defaultProps;
  user: ReturnType<typeof userEvent.setup>;
} {
  const props = { ...defaultProps, ...suppliedProps };
  const user = userEvent.setup();
  const component = render(<MergeModal {...props} />);
  return { ...component, props, user };
}

describe('MergeModal', () => {
  it('does not render modal when showModal is false', () => {
    const { container, queryByTestId } = setup({ showModal: false });

    expect(container).toMatchInlineSnapshot('<div />');
    expect(queryByTestId('confirm-merge')).toBeNull();
  });

  it('renders', async () => {
    const { container, findByTestId, findByText } = setup();

    expect(container).toMatchSnapshot();
    expect(await findByText('Merge "name, name"')).toBeVisible();
    expect(await findByTestId('confirm-merge')).toBeDisabled();
  });

  it('calls handleMergeNameChange when input value changes', async () => {
    const { findByLabelText, props, user } = setup();

    await user.type(await findByLabelText('Name for the merged list'), 'a');

    expect(props.handleMergeNameChange).toHaveBeenCalled();
  });

  it('clears modal when x button is selected', async () => {
    const { findAllByRole, props, user } = setup();

    await user.click((await findAllByRole('button'))[0]);

    expect(props.clearModal).toHaveBeenCalled();
  });

  it('clears modal when Close is selected', async () => {
    const { findByTestId, props, user } = setup();

    await user.click(await findByTestId('clear-merge'));

    expect(props.clearModal).toHaveBeenCalled();
  });

  it('enables confirm when mergeName has a value', async () => {
    const { findByTestId } = setup({ mergeName: 'foo' });

    expect(await findByTestId('confirm-merge')).toBeEnabled();
  });

  it('call handleMergeConfirm when Merge lists is selected', async () => {
    const { findByLabelText, findByTestId, props, user } = setup();

    await user.type(await findByLabelText('Name for the merged list'), 'a');
    await user.click(await findByTestId('confirm-merge'));

    expect(props.handleMergeNameChange).toHaveBeenCalled();
  });

  it('shows warning when lists of different types are selected', async () => {
    const { findByText } = setup({
      selectedLists: [
        { id: '1', name: 'Grocery List', type: EListType.GROCERY_LIST },
        { id: '2', name: 'Book List', type: EListType.BOOK_LIST },
      ],
    });

    expect(await findByText(/Only lists of the same type can be merged/)).toBeVisible();
    expect(await findByText(/while other types will be excluded/)).toBeVisible();
  });

  it('shows detailed breakdown when lists of different types are selected', async () => {
    const { findByText } = setup({
      selectedLists: [
        { id: '1', name: 'Grocery List 1', type: EListType.GROCERY_LIST },
        { id: '2', name: 'Grocery List 2', type: EListType.GROCERY_LIST },
        { id: '3', name: 'Book List', type: EListType.BOOK_LIST },
      ],
    });

    expect(await findByText('Lists to be merged (2):')).toBeVisible();
    expect(await findByText('Lists excluded (1):')).toBeVisible();
    expect(await findByText('Grocery List 1 (GroceryList)')).toBeVisible();
    expect(await findByText('Grocery List 2 (GroceryList)')).toBeVisible();
    expect(await findByText('Book List (BookList)')).toBeVisible();
  });

  it('does not show warning when all lists are the same type', async () => {
    const { queryByText } = setup({
      selectedLists: [
        { id: '1', name: 'Grocery List 1', type: EListType.GROCERY_LIST },
        { id: '2', name: 'Grocery List 2', type: EListType.GROCERY_LIST },
      ],
    });

    expect(queryByText(/Only lists of the same type can be merged/)).toBeNull();
  });
});
