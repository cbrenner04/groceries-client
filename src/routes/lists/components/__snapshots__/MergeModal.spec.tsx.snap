// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeModal renders 1`] = `
<div>
  <div
    class="modal show"
    role="dialog"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        type="button"
      />
      <h5
        class="modal-title"
      >
        Merge 
        "name", "name"
      </h5>
    </div>
    <div
      class="modal-body"
    >
      <form
        class=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="mergeName"
          >
            Name for the merged list
          </label>
          <input
            class="form-control"
            id="mergeName"
            name="mergeName"
            placeholder="My super cool list"
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        data-test-id="clear-merge"
        type="button"
      >
        Close
      </button>
      <button
        class="merge-modal-confirm-disabled btn btn-primary"
        data-test-id="confirm-merge"
        disabled=""
        type="button"
      >
        Merge lists
      </button>
    </div>
  </div>
</div>
`;
